lib/ReactViews/Map/Legend/DistanceLegend.jsx

```diff
@@ -48,179 +47,173 @@ const distances = [
+    const globe = scene.globe;
+    const leftPosition = globe.pick(left, scene);
+    const rightPosition = globe.pick(right, scene);
 
+    if (!defined(leftPosition) || !defined(rightPosition)) {
       this.setState({
-        barWidth: (meters / maxMeters) * maxPixelWidth,
-        distanceLabel: label
+        barWidth: undefined,
+        distanceLabel: undefined
       });
-    },
+      return;
+    }
 
-    render() {
-      const barStyle = {
-        width: this.state.barWidth + "px",
-        left: 5 + (125 - this.state.barWidth) / 2 + "px",
-        height: "2px"
-      };
+    const leftCartographic = globe.ellipsoid.cartesianToCartographic(
+      leftPosition
+    );
+    const rightCartographic = globe.ellipsoid.cartesianToCartographic(
+      rightPosition
+    );
+
+    geodesic.setEndPoints(leftCartographic, rightCartographic);
+    const pixelDistance = geodesic.surfaceDistance;
+
+    // Find the first distance that makes the scale bar less than 100 pixels.
+    const maxBarWidth = 100;
+    let distance;
+    for (let i = distances.length - 1; !defined(distance) && i >= 0; --i) {
+      if (distances[i] / pixelDistance < maxBarWidth) {
+        distance = distances[i];
+      }
+    }
 
-      const distanceLabel = this.state.distanceLabel ? (
-        <div className={Styles.distanceLegend}>
-          <label>{this.state.distanceLabel}</label>
-          <div className={Styles.bar} style={barStyle} />
-        </div>
-      ) : null;
+    if (defined(distance)) {
+      let label;
+      if (distance >= 1000) {
+        label = (distance / 1000).toString() + " km";
+      } else {
+        label = distance.toString() + " m";
+      }
 
-      return distanceLabel;
+      this.setState({
+        barWidth: (distance / pixelDistance) | 0,
+        distanceLabel: label
+      });
+    } else {
+      this.setState({
+        barWidth: undefined,
+        distanceLabel: undefined
+      });
     }
-  })
-);
+  }
+
+  updateDistanceLegendLeaflet(map) {
+    const halfHeight = map.getSize().y / 2;
+    const maxPixelWidth = 100;
+    const maxMeters = map
+      .containerPointToLatLng([0, halfHeight])
+      .distanceTo(map.containerPointToLatLng([maxPixelWidth, halfHeight]));
+
+    const meters = L.control.scale()._getRoundNum(maxMeters);
+    const label = meters < 1000 ? meters + " m" : meters / 1000 + " km";
+
+    this.setState({
+      barWidth: (meters / maxMeters) * maxPixelWidth,
+      distanceLabel: label
+    });
+  }
+
+  render() {
+    const barStyle = {
+      width: this.state.barWidth + "px",
+      left: 5 + (125 - this.state.barWidth) / 2 + "px",
+      height: "2px"
+    };
+
+    const distanceLabel = this.state.distanceLabel ? (
+      <div className={Styles.distanceLegend}>
+        <label>{this.state.distanceLabel}</label>
+        <div className={Styles.bar} style={barStyle} />
+      </div>
+    ) : null;
+
+    return distanceLabel;
+  }
+}
 module.exports = DistanceLegend;
```

Comment:
Did you change this? Or just the spacing?

Reply:
Yes, it is (according to `git diff --ignore-space-change origin/mobx origin/mobx-update-versions -- lib/ReactViews/Map/Legend/DistanceLegend.jsx`)

Reply:
Yeah just a little!

Reply:
You can't use the function version of `disposeOnUnmount` on a `createClass` - needs to be a class component